### PR TITLE
Bring up scheduler benchmark test from lis-test

### PR DESCRIPTION
**Pull request by lubaihua33**
Bring up scheduler benchmark test case from lis-test to LISAv2.
There are two microbenchmarks, hackbench and schbench.
Test results:
```
[LISAv2 Test Results Summary]
Test Run On           : 08/13/2020 11:00:11
ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:22

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 BENCHMARK            PERF-SCHEDULER                                                                    PASS                19.66 
      ARMImageName: Canonical UbuntuServer 18.04-LTS latest, OverrideVMSize: Standard_D16s_v3, TestLocation: westus2, Kernel Version: 5.3.0-1034-azure
      Hackbench test results : =======================================
      HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=14.233
      HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=8.969
      HackbenchType=process.pipe : DataSize_bytes=512 Latency_sec=10.403
      HackbenchType=process.pipe : DataSize_bytes=1024 Latency_sec=12.396
      HackbenchType=thread.socket : DataSize_bytes=2048 Latency_sec=16.862
      HackbenchType=thread.socket : DataSize_bytes=4096 Latency_sec=12.880
      Schbench test results : =======================================
      MessageThreads=6 : Latency95thPercentile_us=35904 Latency99thPercentile_us=53696
      MessageThreads=12 : Latency95thPercentile_us=53696 Latency99thPercentile_us=94080
```

### CODE REVIEW

**Comment by lubaihua33** [comment]
Centos 8
```
[LISAv2 Test Results Summary]
Test Run On           : 08/13/2020 10:30:52
ARM Image Under Test  : OpenLogic : CentOS : 8_1 : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:21

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 BENCHMARK            PERF-SCHEDULER                                                                    PASS                19.29 
      ARMImageName: OpenLogic CentOS 8_1 latest, OverrideVMSize: Standard_D16s_v3, TestLocation: westus2, Kernel Version: 4.18.0-147.8.1.el8_1.x86_64
      Hackbench test results : =======================================
      HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=4.551
      HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=4.847
      HackbenchType=process.pipe : DataSize_bytes=512 Latency_sec=5.395
      HackbenchType=process.pipe : DataSize_bytes=1024 Latency_sec=6.467
      HackbenchType=process.pipe : DataSize_bytes=2048 Latency_sec=8.414
      HackbenchType=process.pipe : DataSize_bytes=4096 Latency_sec=11.248
      HackbenchType=process.pipe : DataSize_bytes=10240 Latency_sec=23.035
      HackbenchType=process.socket : DataSize_bytes=128 Latency_sec=6.653
      HackbenchType=process.socket : DataSize_bytes=256 Latency_sec=7.082
      HackbenchType=process.socket : DataSize_bytes=512 Latency_sec=7.079
      HackbenchType=process.socket : DataSize_bytes=1024 Latency_sec=8.003
      HackbenchType=process.socket : DataSize_bytes=2048 Latency_sec=9.506
      HackbenchType=process.socket : DataSize_bytes=4096 Latency_sec=13.204
      HackbenchType=process.socket : DataSize_bytes=10240 Latency_sec=24.165
      HackbenchType=thread.pipe : DataSize_bytes=128 Latency_sec=4.553
      HackbenchType=thread.pipe : DataSize_bytes=256 Latency_sec=4.826
      HackbenchType=thread.pipe : DataSize_bytes=512 Latency_sec=5.339
      HackbenchType=thread.pipe : DataSize_bytes=1024 Latency_sec=6.223
      HackbenchType=thread.pipe : DataSize_bytes=2048 Latency_sec=7.739
      HackbenchType=thread.pipe : DataSize_bytes=4096 Latency_sec=10.262
      HackbenchType=thread.pipe : DataSize_bytes=10240 Latency_sec=21.205
      HackbenchType=thread.socket : DataSize_bytes=128 Latency_sec=6.593
      HackbenchType=thread.socket : DataSize_bytes=256 Latency_sec=7.111
      HackbenchType=thread.socket : DataSize_bytes=512 Latency_sec=7.048
      HackbenchType=thread.socket : DataSize_bytes=1024 Latency_sec=7.775
      HackbenchType=thread.socket : DataSize_bytes=2048 Latency_sec=9.083
      HackbenchType=thread.socket : DataSize_bytes=4096 Latency_sec=12.660
      HackbenchType=thread.socket : DataSize_bytes=10240 Latency_sec=23.208
      Schbench test results : =======================================
      MessageThreads=6 : Latency95thPercentile_us=58048 Latency99thPercentile_us=87424
      MessageThreads=12 : Latency95thPercentile_us=119168 Latency99thPercentile_us=197376
```

**Comment by lubaihua33** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/13/2020 11:17:43
ARM Image Under Test  : SUSE : sles-15-sp1 : gen1 : 2020.05.27
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:19

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 BENCHMARK            PERF-SCHEDULER                                                                    PASS                17.02 
      ARMImageName: SUSE sles-15-sp1 gen1 2020.05.27, OverrideVMSize: Standard_D16s_v3, TestLocation: westus2, Kernel Version: 4.12.14-8.30-azure
      Hackbench test results : =======================================
      HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=3.709
      HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=3.792
      HackbenchType=thread.socket : DataSize_bytes=2048 Latency_sec=6.687
      HackbenchType=thread.socket : DataSize_bytes=4096 Latency_sec=7.414
      Schbench test results : =======================================
      MessageThreads=6 : Latency95thPercentile_us=56256 Latency99thPercentile_us=91008
      MessageThreads=12 : Latency95thPercentile_us=116096 Latency99thPercentile_us=175872
      ```

File: Testscripts/Linux/perf_scheduler.sh [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+fi
+
+# hackbench
+if [ ! "${DATASIZE}" ]; then
+	DATASIZE=(128 256 512 1024 2048 4096)
+fi
+if [ ! "${LOOPS}" ]; then
+	LOOPS=2000
+fi
+GROUP_NR=$(grep -c ^processor /proc/cpuinfo)
```
Comment (LiliDeng) [inline Testscripts/Linux/perf_scheduler.sh:51]: same with THREADS?

File: Testscripts/Windows/PERF-SCHEDULER.ps1 [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                    $resultMap["KernelVersion"] = $KernelVersion
+                    if ($line[0] -imatch "hackbench") {
+                        $resultMap["TestMode"] = $($Line[0])
+                        $resultMap["DataSize_bytes"] = [Decimal]$($Line[1])
+                        $resultMap["Loops"] = [Decimal]$($Line[2])
+                        $resultMap["Groups"] = [Decimal]$($Line[3])
+                        $resultMap["HackbenchType"] = $($Line[5])
+                        $resultMap["Latency_sec"] = [Decimal]$($Line[6])
+                    } elseif ($line[0] -imatch "schbench") {
+                        $resultMap["TestMode"] = $($Line[0])
```
Comment (LiliDeng) [inline Testscripts/Windows/PERF-SCHEDULER.ps1]: $resultMap["TestMode"] = $($Line[0]) is no difference for schbench and hackbench, move it to common place?

File: Testscripts/Linux/perf_scheduler.sh (hunk 1 above)
Comment (lubaihua33) [reply Testscripts/Linux/perf_scheduler.sh:51]: Yeah.

File: Testscripts/Linux/perf_scheduler.sh [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+function run_hackbench () {
+#   -p, --pipe Sends the data via a pipe instead of the socket (default)
+#   -s, --datasize=<size in bytes> Sets the amount of data to send in each message
+#   -l, --loops=<number of loops> How many messages each sender/receiver pair should send
+#   -g, --groups=<number of groups> Defines how many groups  of  senders  and  receivers  should  be started
+#   -f, --fds=<number of file descriptors> Defines  how  many file descriptors each child should use
+#   -T, --threads Each sender/receiver child will be a POSIX thread of the parent.
+#   -P, --process Hackbench will use fork() on all children (default behaviour)
+	dataSize=${1}
```
Comment (abhimarathe) [inline Testscripts/Linux/perf_scheduler.sh:132]: `dataSize=${1}` Please follow same variable naming throughout the script.

File: XML/TestCases/PerformanceTests.xml [hunk 4]
```diff
@@ -854,6 +854,28 @@
...
+    <testName>PERF-SCHEDULER</testName>
+    <testScript>PERF-SCHEDULER.ps1</testScript>
+    <files>.\Testscripts\Linux\utils.sh,.\Testscripts\Linux\perf_scheduler.sh</files>
+    <TestParameters>
+      <param>MSG_THREADS=SCHBENCH_MSG_THREADS</param>
+      <param>RUNTIME=SCHBENCH_RUNTIME</param>
+      <param>SLEEPTIME=SCHBENCH_SLEEPTIME</param>
+      <param>CPUTIME=SCHBENCH_CPUTIME</param>
+      <param>DATASIZE=HACKBENCH_DATASIZE</param>
+      <param>LOOPS=HACKBENCH_LOOPS</param>
```
Comment (abhimarathe) [inline XML/TestCases/PerformanceTests.xml:867]: Are all of these parameters changed frequently?
If not we can move them inside shell script.

**Review by abhimarathe — COMMENTED** [review]
Please refer to my comments.

**Comment by harz566** [comment]
@lubaihua33   All those Summary text will be save into telemetry database, but if you need further reference those data for some report, suggest leave the last number value for Latency_sec with this format:   xxxxxx,yyy,zzz : 4.551
xxxxxx,yyy,zzz is metadata,  4.551 is the value  in  New-ResultSummary()

HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=4.551
      HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=4.847
      HackbenchType=process.pipe : DataSize_bytes=512 Latency_sec=5.395
      HackbenchType=process.pipe : DataSize_bytes=1024 Latency_sec=6.467
      HackbenchType=process.pipe : DataSize_bytes=2048 Latency_sec=8.414
      HackbenchType=process.pipe : DataSize_bytes=4096 Latency_sec=11.248

File: Testscripts/Windows/PERF-SCHEDULER.ps1 [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            if ($line -imatch "TestMode") {
+                continue;
+            }
+            try {
+                $datasize_bytes = ($line.Trim() -Replace " +"," ").Split(" ")[1]
+                $hackbenchType = ($line.Trim() -Replace " +"," ").Split(" ")[5]
+                $latency_sec = ($line.Trim() -Replace " +"," ").Split(" ")[6]
+                $connResult = "DataSize_bytes=$datasize_bytes Latency_sec=$latency_sec"
+                $metadata = "HackbenchType=$hackbenchType"
+                $currentTestResult.TestSummary += New-ResultSummary -testResult $connResult -metaData $metadata -checkValues "PASS,FAIL,ABORTED" -testName $currentTestData.testName
```
Comment (harz566) [inline Testscripts/Windows/PERF-SCHEDULER.ps1:69]: All those Summary text will be save into telemetry database, but if you need further reference those data for some report, suggest leave the last number value for Latency_sec with this format:   xxxxxx,yyy,zzz : 4.551
xxxxxx,yyy,zzz is metadata,  4.551 is the value  in  New-ResultSummary()

File: Testscripts/Linux/perf_scheduler.sh [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+	fi
+}
+
+function build_hackbench () {
+	pkgver=1.8
+	source="https://www.kernel.org/pub/linux/utils/rt-tests/rt-tests-${pkgver}.tar.gz"
+	wget $source
+	tar -zxvf rt-tests-${pkgver}.tar.gz
+	cd rt-tests-${pkgver}
+	make
```
Comment (johnsongeorge-w) [inline Testscripts/Linux/perf_scheduler.sh]: How about ```make && make install```. This will allow to check the successful execution of both commands.

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]

File: Testscripts/Windows/PERF-SCHEDULER.ps1 [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+. utils.sh
+collect_VM_properties
+"@
+        Set-Content "$LogDir\Start${testName}Test.sh" $myString
+        Copy-RemoteFiles -uploadTo $AllVMData.PublicIP -port $AllVMData.SSHPort `
+            -files "$constantsFile,$LogDir\Start${testName}Test.sh" -username $user -password $password -upload
+        Copy-RemoteFiles -uploadTo $AllVMData.PublicIP -port $AllVMData.SSHPort `
+            -files $currentTestData.files -username $user -password $password -upload
+
+        Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password `
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/PERF-SCHEDULER.ps1:31]: @johnsongeorge-w recommended me to use "bash \<bash script\>" without file mode change. Try it!

File: Testscripts/Windows/PERF-SCHEDULER.ps1 [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            -command "chmod +x *.sh" -runAsSudo | Out-Null
+        $testJob = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password `
+            -command "./Start${testName}Test.sh" -RunInBackground -runAsSudo
+
+        while ((Get-Job -Id $testJob).State -eq "Running") {
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/PERF-SCHEDULER.ps1:36]: What is the exit condition of this while loop?

File: Testscripts/Windows/PERF-SCHEDULER.ps1 [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+            $currentStatus = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password `
+                -command "tail -2 ${testName}ConsoleLogs.txt | head -1"
+            Write-LogInfo "Current Test Status : $currentStatus"
+            Wait-Time -seconds 20
+        }
+        $finalStatus = Run-LinuxCmd -ip $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password `
+            -command "cat ./state.txt"
+        Copy-RemoteFiles -downloadFrom $AllVMData.PublicIP -port $AllVMData.SSHPort -username $user -password $password `
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/PERF-SCHEDULER.ps1]: Keep them in the loop form.

**Review by juhlee-microsoft — CHANGES_REQUESTED** [review]

File: XML/TestCases/PerformanceTests.xml (hunk 4 above)
Comment (lubaihua33) [reply XML/TestCases/PerformanceTests.xml:867]: Not frequently. But these parameters can be defined by users, so I put them here.

**Comment by lubaihua33** [comment]
> @lubaihua33 All those Summary text will be save into telemetry database, but if you need further reference those data for some report, suggest leave the last number value for Latency_sec with this format: xxxxxx,yyy,zzz : 4.551
> xxxxxx,yyy,zzz is metadata, 4.551 is the value in New-ResultSummary()
> 
> HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=4.551
> HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=4.847
> HackbenchType=process.pipe : DataSize_bytes=512 Latency_sec=5.395
> HackbenchType=process.pipe : DataSize_bytes=1024 Latency_sec=6.467
> HackbenchType=process.pipe : DataSize_bytes=2048 Latency_sec=8.414
> HackbenchType=process.pipe : DataSize_bytes=4096 Latency_sec=11.24

Thanks for your suggestion.

**Comment by lubaihua33** [comment]
> @lubaihua33 All those Summary text will be save into telemetry database, but if you need further reference those data for some report, suggest leave the last number value for Latency_sec with this format: xxxxxx,yyy,zzz : 4.551
> xxxxxx,yyy,zzz is metadata, 4.551 is the value in New-ResultSummary()
> 
> HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=4.551
> HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=4.847
> HackbenchType=process.pipe : DataSize_bytes=512 Latency_sec=5.395
> HackbenchType=process.pipe : DataSize_bytes=1024 Latency_sec=6.467
> HackbenchType=process.pipe : DataSize_bytes=2048 Latency_sec=8.414
> HackbenchType=process.pipe : DataSize_bytes=4096 Latency_sec=11.248

I'm not sure clearly. If xxxxxx,yyy,zzz is metadata, 4.551 is the value in New-ResultSummary(), the output format is xxxxxx,yyy,zzz =4.551.

File: Testscripts/Windows/PERF-SCHEDULER.ps1 (hunk 5 above)
Comment (lubaihua33) [reply Testscripts/Windows/PERF-SCHEDULER.ps1:69]: If xxxxxx,yyy,zzz is metadata, 4.551 is the value in New-ResultSummary(), the output format is xxxxxx,yyy,zzz =4.551. I think this format is good.

File: Testscripts/Windows/PERF-SCHEDULER.ps1 (hunk 8 above)
Comment (lubaihua33) [reply Testscripts/Windows/PERF-SCHEDULER.ps1:36]: The exit condition is that the $testJob is completed, which is the state of $testJob is not running.

File: Testscripts/Windows/PERF-SCHEDULER.ps1 (hunk 7 above)
Comment (lubaihua33) [reply Testscripts/Windows/PERF-SCHEDULER.ps1:31]: I had tried. But I also need to "chmod utils.sh", or there will be "Permission denied" error.

**Comment by lubaihua33** [comment]
```
[LISAv2 Test Results Summary]
Test Run On           : 08/24/2020 02:35:20
ARM Image Under Test  : OpenLogic : CentOS : 8_1 : latest
Total Test Cases      : 1 (1 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:19

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 BENCHMARK            PERF-SCHEDULER                                                                    PASS                17.09 
      ARMImageName: OpenLogic CentOS 8_1 latest, OverrideVMSize: Standard_D16s_v3, TestLocation: westus2, Kernel Version: 4.18.0-147.8.1.el8_1.x86_64
      Hackbench test results : =======================================
      HackbenchType=process.pipe : DataSize_bytes=128 Latency_sec=4.456
      HackbenchType=process.pipe : DataSize_bytes=256 Latency_sec=4.741
      HackbenchType=process.pipe : DataSize_bytes=512 Latency_sec=5.267
      HackbenchType=process.pipe : DataSize_bytes=1024 Latency_sec=6.247
      HackbenchType=process.pipe : DataSize_bytes=2048 Latency_sec=7.796
      HackbenchType=process.pipe : DataSize_bytes=4096 Latency_sec=10.470
      HackbenchType=process.socket : DataSize_bytes=128 Latency_sec=6.310
      HackbenchType=process.socket : DataSize_bytes=256 Latency_sec=6.499
      HackbenchType=process.socket : DataSize_bytes=512 Latency_sec=6.634
      HackbenchType=process.socket : DataSize_bytes=1024 Latency_sec=7.235
      HackbenchType=process.socket : DataSize_bytes=2048 Latency_sec=8.564
      HackbenchType=process.socket : DataSize_bytes=4096 Latency_sec=12.424
      Schbench test results : =======================================
      MessageThreads=6 : Latency95thPercentile_us=57152 Latency99thPercentile_us=86912
      MessageThreads=12 : Latency95thPercentile_us=117888 Latency99thPercentile_us=195328
```

File: Testscripts/Windows/PERF-SCHEDULER.ps1 (hunk 7 above)
Comment (johnsongeorge-w) [reply Testscripts/Windows/PERF-SCHEDULER.ps1:31]: Can you please give more details. I think utils.sh is sourced in test script and there should not be any issue when running using bash.

**Review by juhlee-microsoft — DISMISSED** [review]

File: Testscripts/Linux/perf_scheduler.sh [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+function start_monitor () {
+	name=${1}
+	type=${2}
+	vmstat 1 2>&1 > /${LOG_FOLDER}/${name}.${type}.vmstat.log 2>&1 & 
+	sar -P ALL 1 2>&1 > /${LOG_FOLDER}/${name}.${type}.sar.cpu.log 2>&1 &
+}
+
+function stop_monitor () {
+    pkill -f vmstat
```
Comment (LiliDeng) [inline Testscripts/Linux/perf_scheduler.sh]: keep the same indent

**Review by harz566 — APPROVED** [review]

**Review by LiliDeng — APPROVED** [review]

**Review by johnsongeorge-w — APPROVED** [review]

**Pull request merged**